lib/kafka_producer.js

```diff
@@ -91,18 +110,82 @@ KafkaProducer.prototype.connect = function connect(onConnect) {
 
 KafkaProducer.prototype.produce = function produce(topic, message, timeStamp, callback) {
     var self = this;
+
     if (self.restClient) {
-        var produceMessage = self.getProduceMessage(topic, message, timeStamp, 'binary');
-        self.restClient.produce(produceMessage, function handleResponse(err, res) {
-            if (callback) {
-                callback(err, res);
-            }
-        });
+        if (self.batching) {
+            self.batch(topic, message, timeStamp, callback);
+        } else {
+            var produceMessage = self.getProduceMessage(topic, message, timeStamp, 'binary');
+            self.restClient.produce(produceMessage, function handleResponse(err, res) {
+                if (callback) {
+                    callback(err, res);
+                }
+            });
+        }
     } else if (callback) {
         callback(new Error('Kafka Rest Client is not initialized!'));
     }
 };
 
+// Add message to topic's BatchMessage in cache or add new topic to cache
+// Flush topic's BatchMessage on flushCycleSecs interval or if greater than maxBatchSizeBytes
+KafkaProducer.prototype.batch = function batch(topic, message, timeStamp, callback) {
+    var self = this;
+    var messageBatch;
+
+    if (self.topicToBatchQueue[topic]) {
+        messageBatch = self.topicToBatchQueue[topic];
+        if ((messageBatch.sizeBytes + message.length) > self.maxBatchSizeBytes) {
+            var produceMessage = self.getProduceMessage(topic,
+                messageBatch.getBatchedMessage(),
+                timeStamp,
+                'batch');
+            self.restClient.produce(produceMessage, function handleResponse(err, res) {
+                if (callback) {
+                    callback(err, res);
+                }
+            });
+
+            self.topicToBatchQueue[topic] = new MessageBatch(self.maxBatchSizeBytes);
```

Comment:
You cant do this. Every message batch allocates a huge buffer. Either cache MessageBatch or cache buffer.